File: android/app/src/main/java/com/flow/app/Protocol.kt
package com.flow.app

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject

/**
 * Wire protocol & data contracts — Kotlin side.
 *
 * This file is the Android mirror of `shared/protocol.md` and MUST stay byte-for-byte
 * wire-compatible with the desktop engine. Every JSON field name here is taken
 * verbatim from `shared/protocol.md`; do not rename without updating the spec and
 * the desktop implementation in lock-step.
 *
 * Constants mirror `shared/config.json`. Keep [Config] in sync with that file.
 */
object Config {
    const val PROTOCOL_VERSION = 1
    const val APP_NAME = "Flow"

    const val TEXT_DIM = 384      // all-MiniLM-L6-v2 (text/semantic space)
    const val IMAGE_DIM = 512     // OpenCLIP ViT-B/32 (image/zero-shot space)
    const val WS_PORT = 8787      // federation websocket
    const val TOP_K = 8           // per-node retrieval
    const val QUERY_TIMEOUT_MS = 1500L

    val SOURCES = listOf("trove", "trail", "files", "sieve", "relay", "threads", "audio")
    val TYPES = listOf(
        "wifi", "parking", "receipt", "serial", "poster",
        "event", "doc", "activity", "note", "contact", "other"
    )

    object Models {
        const val TEXT_EMBED = "sentence-transformers/all-MiniLM-L6-v2"
        const val TEXT_EMBED_DIM = 384
        const val IMAGE_EMBED = "ViT-B-32"
        const val IMAGE_EMBED_DIM = 512
        const val LLM_PHONE = "Llama-3.2-3B-Instruct"
        const val LLM_DESKTOP = "Llama-3.1-8B-Instruct"
    }

    object Crypto {
        const val KDF = "HKDF-SHA256"
        const val AEAD = "XChaCha20-Poly1305"
        const val PSK_BYTES = 32
    }
}

/** Shared JSON codec. `encodeDefaults` keeps `v` on the wire; `ignoreUnknownKeys` is forward-compatible. */
val FlowJson: Json = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
    explicitNulls = true
}

/**
 * Item — the unit stored in every node's local index (see shared/protocol.md §Item).
 *
 * `fields` is an arbitrary structured map of extracted values; we keep it as a JSON object
 * so it round-trips losslessly with the desktop engine regardless of the keys present.
 */
@Serializable
data class Item(
    val id: String,
    val device_id: String,
    val source: String,                 // one of Config.SOURCES
    val ts: Long,                       // unix seconds
    val app_context: String? = null,    // "IRCTC" | "Chrome/proposal.docx" | null
    val text: String,                   // extracted / ocr / transcript text
    val type: String,                   // one of Config.TYPES
    val fields: JsonObject = JsonObject(emptyMap()),
    val thumb_b64: String? = null,      // base64 jpeg <= 64KB | null
    val file_ref: String? = null        // origin-device path for fetch-on-open | null
)

/** Hit — a single search match returned in RESULTS (see shared/protocol.md §Hit). */
@Serializable
data class Hit(
    val item_id: String,
    val device_id: String,
    val score: Double,
    val source: String,
    val type: String,
    val text: String,                   // snippet
    val fields: JsonObject = JsonObject(emptyMap()),
    val thumb_b64: String? = null
)

/**
 * Capabilities advertised in HELLO.
 * `tops` = NPU TOPS, `has_llm` = can generate locally, `battery` = 0..100 (or -1 if plugged/unknown).
 */
@Serializable
data class Caps(
    val tops: Double = 0.0,
    val has_llm: Boolean = false,
    val battery: Int = -1
)

// ---------------------------------------------------------------------------
// Messages. Every message is { "type": <string>, "v": 1, ...payload }.
// We model the envelope discriminator as a sealed hierarchy on `type`.
// ---------------------------------------------------------------------------

/** Names of the message `type` discriminator — keep identical to shared/protocol.md. */
object MsgType {
    const val HELLO = "HELLO"
    const val QUERY = "QUERY"
    const val RESULTS = "RESULTS"
    const val FETCH = "FETCH"
    const val FETCH_RESULT = "FETCH_RESULT"
    const val SURFACE = "SURFACE"
}

@Serializable
data class Hello(
    val type: String = MsgType.HELLO,
    val v: Int = Config.PROTOCOL_VERSION,
    val device_id: String,
    // Defaulted for parity with the desktop, which models caps as a free-form dict and may
    // emit caps={} or omit it. ignoreUnknownKeys + Caps' own member defaults keep decode tolerant.
    val name: String = "",
    val caps: Caps = Caps()
)

@Serializable
data class Query(
    val type: String = MsgType.QUERY,
    val v: Int = Config.PROTOCOL_VERSION,
    val query_id: String,
    val text: String,
    val top_k: Int = Config.TOP_K
)

@Serializable
data class Results(
    val type: String = MsgType.RESULTS,
    val v: Int = Config.PROTOCOL_VERSION,
    val query_id: String,
    val device_id: String,
    val hits: List<Hit> = emptyList()
)

@Serializable
data class Fetch(
    val type: String = MsgType.FETCH,
    val v: Int = Config.PROTOCOL_VERSION,
    val item_id: String
)

@Serializable
data class FetchResult(
    val type: String = MsgType.FETCH_RESULT,
    val v: Int = Config.PROTOCOL_VERSION,
    val item_id: String,
    val mime: String = "application/octet-stream",
    val blob_b64: String = ""
)

@Serializable
data class Surface(
    val type: String = MsgType.SURFACE,
    val v: Int = Config.PROTOCOL_VERSION,
    val event: String,
    // Constrained to a JSON object to match the desktop's `dict(d.get("payload") or {})`
    // handling (shared/protocol.md shows {event, payload}); a non-object payload would
    // fail to decode on the Python side, so Android only ever emits an object here.
    val payload: JsonObject = JsonObject(emptyMap())
)

/**
 * Pairing QR payload (shared/protocol.md §Pairing):
 *   { "ip": "192.168.1.20", "port": 8787, "psk": "<base64 32-byte key>", "v": 1 }
 */
@Serializable
data class PairingInfo(
    val ip: String,
    val port: Int = Config.WS_PORT,
    val psk: String,                    // base64 of 32 raw bytes
    val v: Int = Config.PROTOCOL_VERSION
)

/**
 * Lightweight, allocation-free way to read just the `type`/`v` of an inbound frame
 * before deciding which data class to decode it into. Returns null if not JSON or no type.
 */
object Wire {
    fun peekType(jsonText: String): String? = try {
        val obj = FlowJson.parseToJsonElement(jsonText) as? JsonObject ?: return null
        (obj["type"] as? kotlinx.serialization.json.JsonPrimitive)?.content
    } catch (_: Throwable) {
        null
    }

    fun encode(item: Item): String = FlowJson.encodeToString(Item.serializer(), item)
    fun decodeItem(s: String): Item = FlowJson.decodeFromString(Item.serializer(), s)
}

/** The asker's final composed answer (shared/protocol.md §Query → answer flow, step 6). */
@Serializable
data class AnswerSource(
    val item_id: String,
    val device_id: String,
    val file_ref: String? = null
)

@Serializable
data class Answer(
    val answer: String,
    val sources: List<AnswerSource>
)

File: desktop/flow/protocol.py
"""Wire protocol — dataclasses for Item, Hit and every message type.

Field names match ``shared/protocol.md`` EXACTLY so the desktop engine and the
Android app are wire-compatible. Every message is a JSON object of the form::

    { "type": <string>, "v": 1, ...payload }

and is carried inside the AEAD-encrypted frame (see ``crypto`` and ``federation``).
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field, asdict
from typing import Any, Dict, List, Optional

from .config import PROTOCOL_VERSION, TOP_K

# ---------------------------------------------------------------------------
# Message type constants
# ---------------------------------------------------------------------------
HELLO = "HELLO"
QUERY = "QUERY"
RESULTS = "RESULTS"
FETCH = "FETCH"
FETCH_RESULT = "FETCH_RESULT"
SURFACE = "SURFACE"

MESSAGE_TYPES = {HELLO, QUERY, RESULTS, FETCH, FETCH_RESULT, SURFACE}


def new_id() -> str:
    """Generate a fresh UUID4 string (used for item ids and query ids)."""
    return str(uuid.uuid4())


# ---------------------------------------------------------------------------
# Core records
# ---------------------------------------------------------------------------
@dataclass
class Item:
    """The unit stored in every node's local index (see protocol.md §Item)."""

    id: str
    device_id: str
    source: str            # trove|trail|files|sieve|relay|threads|audio
    ts: int                # unix seconds
    text: str
    type: str              # wifi|parking|receipt|serial|poster|event|doc|activity|note|contact|other
    app_context: Optional[str] = None
    fields: Dict[str, Any] = field(default_factory=dict)
    thumb_b64: Optional[str] = None
    file_ref: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "device_id": self.device_id,
            "source": self.source,
            "ts": int(self.ts),
            "app_context": self.app_context,
            "text": self.text,
            "type": self.type,
            "fields": dict(self.fields or {}),
            "thumb_b64": self.thumb_b64,
            "file_ref": self.file_ref,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Item":
        return cls(
            id=d.get("id") or new_id(),
            device_id=d["device_id"],
            source=d["source"],
            ts=int(d.get("ts", 0)),
            app_context=d.get("app_context"),
            text=d.get("text", "") or "",
            type=d.get("type", "other") or "other",
            fields=dict(d.get("fields") or {}),
            thumb_b64=d.get("thumb_b64"),
            file_ref=d.get("file_ref"),
        )


@dataclass
class Hit:
    """A single match returned in answer to a QUERY (see protocol.md §Hit)."""

    item_id: str
    device_id: str
    score: float
    source: str
    type: str
    text: str
    fields: Dict[str, Any] = field(default_factory=dict)
    thumb_b64: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "item_id": self.item_id,
            "device_id": self.device_id,
            "score": float(self.score),
            "source": self.source,
            "type": self.type,
            "text": self.text,
            "fields": dict(self.fields or {}),
            "thumb_b64": self.thumb_b64,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Hit":
        return cls(
            item_id=d["item_id"],
            device_id=d["device_id"],
            score=float(d.get("score", 0.0)),
            source=d.get("source", "other"),
            type=d.get("type", "other"),
            text=d.get("text", "") or "",
            fields=dict(d.get("fields") or {}),
            thumb_b64=d.get("thumb_b64"),
        )

    @classmethod
    def from_item(cls, item: Item, score: float) -> "Hit":
        return cls(
            item_id=item.id,
            device_id=item.device_id,
            score=float(score),
            source=item.source,
            type=item.type,
            text=item.text,
            fields=dict(item.fields or {}),
            thumb_b64=item.thumb_b64,
        )


# ---------------------------------------------------------------------------
# Messages
# ---------------------------------------------------------------------------
@dataclass
class Hello:
    """Post-pair handshake. ``caps`` = {tops, has_llm, battery}."""

    device_id: str
    name: str
    caps: Dict[str, Any] = field(default_factory=dict)
    v: int = PROTOCOL_VERSION

    def to_dict(self) -> Dict[str, Any]:
        return {"type": HELLO, "v": self.v, "device_id": self.device_id,
                "name": self.name, "caps": dict(self.caps or {})}

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Hello":
        return cls(device_id=d["device_id"], name=d.get("name", ""),
                   caps=dict(d.get("caps") or {}), v=int(d.get("v", PROTOCOL_VERSION)))


@dataclass
class Query:
    """Federated search request, A -> peers."""

    query_id: str
    text: str
    top_k: int = TOP_K
    v: int = PROTOCOL_VERSION

    def to_dict(self) -> Dict[str, Any]:
        return {"type": QUERY, "v": self.v, "query_id": self.query_id,
                "text": self.text, "top_k": int(self.top_k)}

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Query":
        return cls(query_id=d["query_id"], text=d.get("text", ""),
                   top_k=int(d.get("top_k", TOP_K)), v=int(d.get("v", PROTOCOL_VERSION)))


@dataclass
class Results:
    """Per-node top-k matches, peer -> A."""

    query_id: str
    device_id: str
    hits: List[Hit] = field(default_factory=list)
    v: int = PROTOCOL_VERSION

    def to_dict(self) -> Dict[str, Any]:
        return {"type": RESULTS, "v": self.v, "query_id": self.query_id,
                "device_id": self.device_id, "hits": [h.to_dict() for h in self.hits]}

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Results":
        return cls(query_id=d["query_id"], device_id=d["device_id"],
                   hits=[Hit.from_dict(h) for h in d.get("hits", [])],
                   v=int(d.get("v", PROTOCOL_VERSION)))


@dataclass
class Fetch:
    """Request the full file for an opened result, A -> peer."""

    item_id: str
    v: int = PROTOCOL_VERSION

    def to_dict(self) -> Dict[str, Any]:
        return {"type": FETCH, "v": self.v, "item_id": self.item_id}

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Fetch":
        return cls(item_id=d["item_id"], v=int(d.get("v", PROTOCOL_VERSION)))


@dataclass
class FetchResult:
    """The full file, peer -> A."""

    item_id: str
    mime: str
    blob_b64: str
    v: int = PROTOCOL_VERSION

    def to_dict(self) -> Dict[str, Any]:
        return {"type": FETCH_RESULT, "v": self.v, "item_id": self.item_id,
                "mime": self.mime, "blob_b64": self.blob_b64}

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "FetchResult":
        return cls(item_id=d["item_id"], mime=d.get("mime", "application/octet-stream"),
                   blob_b64=d.get("blob_b64", ""), v=int(d.get("v", PROTOCOL_VERSION)))


@dataclass
class Surface:
    """Proactive recall notification (push)."""

    event: str
    payload: Dict[str, Any] = field(default_factory=dict)
    v: int = PROTOCOL_VERSION

    def to_dict(self) -> Dict[str, Any]:
        return {"type": SURFACE, "v": self.v, "event": self.event,
                "payload": dict(self.payload or {})}

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Surface":
        # protocol.md intends ``payload`` to be a JSON object. The Android side
        # types it as a generic JsonElement, so a peer *could* emit a non-object
        # (array/scalar/null). Degrade gracefully instead of raising on decode:
        # coerce any non-object payload into a wrapped object so the core path
        # never crashes (see protocol findings — SURFACE payload width mismatch).
        raw = d.get("payload")
        if isinstance(raw, dict):
            payload: Dict[str, Any] = dict(raw)
        elif raw is None:
            payload = {}
        else:
            payload = {"value": raw}
        return cls(event=d["event"], payload=payload,
                   v=int(d.get("v", PROTOCOL_VERSION)))


# ---------------------------------------------------------------------------
# Generic (de)serialization helper
# ---------------------------------------------------------------------------
_DECODERS = {
    HELLO: Hello.from_dict,
    QUERY: Query.from_dict,
    RESULTS: Results.from_dict,
    FETCH: Fetch.from_dict,
    FETCH_RESULT: FetchResult.from_dict,
    SURFACE: Surface.from_dict,
}


def decode_message(d: Dict[str, Any]):
    """Decode a JSON dict into the appropriate message dataclass.

    Raises ``ValueError`` for unknown types or major-version mismatch.
    """
    mtype = d.get("type")
    if mtype not in _DECODERS:
        raise ValueError(f"unknown message type: {mtype!r}")
    v = int(d.get("v", PROTOCOL_VERSION))
    if v != PROTOCOL_VERSION:
        # Mismatched majors must refuse (protocol.md §Versioning).
        raise ValueError(f"protocol version mismatch: got {v}, expected {PROTOCOL_VERSION}")
    return _DECODERS[mtype](d)
